presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveBucketing.java

```diff
@@ -126,6 +144,54 @@ public static int getHiveBucket(BucketingVersion bucketingVersion, int bucketCou
         return getBucketNumber(bucketingVersion.getBucketHashCode(types, values), bucketCount);
     }
 
+    private static Optional<Set<Integer>> getHiveBuckets(BucketingVersion bucketingVersion, int bucketCount, List<TypeInfo> types, List<List<NullableValue>> values)
+    {
+        long explorationCount;
+        try {
+            explorationCount = values.stream()
+                    .mapToLong(List::size)
+                    .reduce(1, Math::multiplyExact);
+        }
+        catch (ArithmeticException e) {
+            return Optional.empty();
+        }
+        long explorationLimit = Math.max(bucketCount * BUCKETS_EXPLORATION_LIMIT_FACTOR, BUCKETS_EXPLORATION_GRANTED_LIMIT);
+        if (explorationCount > explorationLimit) {
+            return Optional.empty();
+        }
+
+        Set<Integer> buckets = new HashSet<>();
+        getHiveBucketsByDiscreteSets(bucketingVersion, bucketCount, types, values, buckets, 0, 0);
+        if (buckets.size() < bucketCount) {
+            return Optional.of(ImmutableSet.copyOf(buckets));
+        }
+        return Optional.empty();
+    }
+
+    private static void getHiveBucketsByDiscreteSets(
```

Comment:
A non-recursive implementation would be easier to read, IMO (you'll need to make `values` a list of sets, though):

```java
return Sets.cartesianProduct(values).stream()
        .map(entry -> bucketingVersion.getBucketHashCode(types, entry.stream().map(NullableValue::getValue).toArray()))
        .collect(Collectors.toSet());
```

(BTW, this method becomes unnecessary with this simplification -- just inline it in the caller)

Reply:
I used the Cartesian product, but used a for-loop instead of stream. This way I can break the bucket enumeration when I find that all buckets were already hit.
BTW you can have a Cartesian product on lists, too.